File: myExpenses/src/main/java/org/totschnig/myexpenses/util/crashreporting/CrashHandler.java
package org.totschnig.myexpenses.util.crashreporting;

import android.content.Context;

import org.totschnig.myexpenses.MyApplication;
import org.totschnig.myexpenses.preference.PrefKey;
import org.totschnig.myexpenses.provider.DbUtils;
import org.totschnig.myexpenses.util.distrib.DistributionHelper;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import timber.log.Timber;

public abstract class CrashHandler {
  private static final String CUSTOM_DATA_KEY_BREADCRUMB = "Breadcrumb";
  private String currentBreadCrumb;
  private boolean enabled;

  public static void reportWithDbSchema(Exception e) {
    report(e, DbUtils.getSchemaDetails());
  }

  public static void report(Exception e, Map<String, String> customData) {
    for (Map.Entry<String, String> entry : customData.entrySet()) {
      Timber.w("%s: %s", entry.getKey(), entry.getValue());
    }
    report(e);
  }

  public static void report(Exception e, String key, String data) {
    Timber.w("%s: %s", key, data);
    report(e);
  }

  public static void report(Throwable e) {
    report(e, null);
  }

  public static void report(Throwable e, String tag) {
    if (tag != null) {
      Timber.tag(tag);
    }
    Timber.e(e);
  }

  public static void reportWithTag(String message, String tag) {
    final Exception e = new Exception(message);
    e.fillInStackTrace();
    List<StackTraceElement> stack = new ArrayList<>(Arrays.asList(e.getStackTrace()));
    stack.remove(0);
    e.setStackTrace(stack.toArray(new StackTraceElement[0]));
    report(e, tag);
  }

  public static void reportWithFormat(String format, Object... args) {
    final Exception e = new Exception(String.format(format, args));
    e.fillInStackTrace();
    List<StackTraceElement> stack = new ArrayList<>(Arrays.asList(e.getStackTrace()));
    stack.remove(0);
    e.setStackTrace(stack.toArray(new StackTraceElement[0]));
    report(e);
  }

  public static void report(String message) {
    final Exception e = new Exception(message);
    e.fillInStackTrace();
    List<StackTraceElement> stack = new ArrayList<>(Arrays.asList(e.getStackTrace()));
    stack.remove(0);
    e.setStackTrace(stack.toArray(new StackTraceElement[0]));
    report(e);
  }

  public abstract void onAttachBaseContext(MyApplication application);

  public void setupLogging(Context context) {
    enabled = PrefKey.CRASHREPORT_ENABLED.getBoolean(true);
    if (enabled) {
      setupLoggingDo(context);
    }
  }

  protected void setKeys(Context context) {
    putCustomData("Distribution", DistributionHelper.getVersionInfo(context));
    putCustomData("Installer", context.getPackageManager().getInstallerPackageName(context.getPackageName()));
    putCustomData("Locale", Locale.getDefault().toString());
  }

  abstract void setupLoggingDo(Context context);

  public void setUserEmail(String value) {
    putCustomData("UserEmail", value);
  }

  public abstract void putCustomData(@NonNull String key, @Nullable String value);

  public synchronized void addBreadcrumb(String breadcrumb) {
    Timber.i("Breadcrumb: %s", breadcrumb);
    if (enabled) {
      currentBreadCrumb = currentBreadCrumb == null ? "" : currentBreadCrumb.substring(Math.max(0, currentBreadCrumb.length() - 500));
      currentBreadCrumb += "->" + breadcrumb;
      putCustomData(CUSTOM_DATA_KEY_BREADCRUMB, currentBreadCrumb);
    }
  }

  public static CrashHandler NO_OP = new CrashHandler() {
    @Override
    public void onAttachBaseContext(MyApplication application) {

    }

    @Override
    void setupLoggingDo(Context context) {

    }

    @Override
    public void putCustomData(@NonNull String key, String value) {

    }
  };

  public void initProcess(Context context, boolean syncService) {}
}

File: myExpenses/src/main/java/org/totschnig/myexpenses/util/crashreporting/CrashHandler.kt
package org.totschnig.myexpenses.util.crashreporting

import android.content.ContentResolver
import android.content.Context
import org.totschnig.myexpenses.BuildConfig
import org.totschnig.myexpenses.MyApplication
import org.totschnig.myexpenses.preference.PrefHandler
import org.totschnig.myexpenses.preference.PrefKey
import org.totschnig.myexpenses.provider.DbUtils
import org.totschnig.myexpenses.util.distrib.DistributionHelper.getVersionInfo
import timber.log.Timber
import java.util.*

interface CrashHandler {
    fun onAttachBaseContext(application: MyApplication) {}
    suspend fun setupLogging(context: Context) {}
    fun putCustomData(key: String, value: String) {}
    fun setEnabled(enabled: Boolean) {}
    fun setUserEmail(email: String?) {}
    fun addBreadcrumb(breadcrumb: String) {}
    fun initProcess(context: Context, syncService: Boolean) {}
    fun  getInfo(): Pair<String, String>? = null

    companion object {
        @JvmStatic
        fun reportWithDbSchema(contentResolver: ContentResolver, e: Throwable) {
            report(e, DbUtils.getSchemaDetails(contentResolver))
        }

        @JvmStatic
        @JvmOverloads
        fun report(e: Throwable, customData: Map<String, String>, tag: String? = null) {
            for ((key, value) in customData) {
                tag?.let { Timber.tag(it) }
                Timber.w("%s: %s", key, value)
            }
            report(e, tag)
        }

        @JvmStatic
        @JvmOverloads
        fun report(e: Throwable, key: String, data: String, tag: String? = null) {
            tag?.let { Timber.tag(it) }
            Timber.w("%s: %s", key, data)
            report(e, tag)
        }

        @JvmStatic
        @JvmOverloads
        fun throwOrReport(message: String, tag: String? = null) {
            throwOrReport(IllegalStateException(message), tag)
        }

        fun throwOrReport(e: Throwable, tag: String? = null) {
            if (BuildConfig.DEBUG) {
                throw e
            } else {
                report(e, tag)
            }
        }

        @JvmStatic
        @JvmOverloads
        fun report(e: Throwable, tag: String? = null) {
            tag?.let { Timber.tag(it) }
            Timber.e(e)
        }

        val NO_OP: CrashHandler = object : CrashHandler {}
    }
}

abstract class BaseCrashHandler(val prefHandler: PrefHandler): CrashHandler {
    private var currentBreadCrumb: String? = null

    open fun setKeys(context: Context) {
        putCustomData("Distribution", getVersionInfo(context))
        context.packageManager.getInstallerPackageName(context.packageName)?.let {
            putCustomData("Installer", it)
        }
        putCustomData("Locale", Locale.getDefault().toString())
        putCustomData("Protection", when {
            prefHandler.getBoolean(PrefKey.PROTECTION_LEGACY, false) -> "Legacy"
            prefHandler.getBoolean(PrefKey.PROTECTION_DEVICE_LOCK_SCREEN, false) -> "Device"
            else -> "None"
        })
    }

    override fun setUserEmail(email: String?) {
        if (email != null) {
            putCustomData("UserEmail", email)
        }
    }

    @Synchronized
    override fun addBreadcrumb(breadcrumb: String) {
        Timber.i("Breadcrumb: %s", breadcrumb)
        currentBreadCrumb =
            (if (currentBreadCrumb == null) "" else currentBreadCrumb!!.substring(
                0.coerceAtLeast(currentBreadCrumb!!.length - 500)
            ) + "->$breadcrumb").also {
                putCustomData(CUSTOM_DATA_KEY_BREADCRUMB, it)
            }
    }

    companion object {
        private const val CUSTOM_DATA_KEY_BREADCRUMB = "Breadcrumb"
    }

}